File: 2 Data Structures/Week 6 Binary Search Trees 2/Programming-Assignment-4/tree_orders/tree-orders.py
# python3

import sys, threading
sys.setrecursionlimit(10**6) # max depth of recursion
threading.stack_size(2**27)  # new thread will get stack of such size

class TreeOrders:
  def read(self):
    self.n = int(sys.stdin.readline())
    self.key = [0 for i in range(self.n)]
    self.left = [0 for i in range(self.n)]
    self.right = [0 for i in range(self.n)]
    for i in range(self.n):
      [a, b, c] = map(int, sys.stdin.readline().split())
      self.key[i] = a
      self.left[i] = b
      self.right[i] = c

  def inOrder(self):
    self.result = []
    # Finish the implementation
    # You may need to add a new recursive method to do that
    def inOrderTraversal(i, result):
        if self.left[i] != -1:
            inOrderTraversal(self.left[i], result)
        result.append(self.key[i])
        if self.right[i] != -1:
            inOrderTraversal(self.right[i], result)
    inOrderTraversal(0, self.result)
                
    return self.result

  def preOrder(self):
    self.result = []
    # Finish the implementation
    # You may need to add a new recursive method to do that
    def preOrderTraversal(i, result):
        result.append(self.key[i])
        if self.left[i] != -1:
            preOrderTraversal(self.left[i], result)        
        if self.right[i] != -1:
            preOrderTraversal(self.right[i], result)
    preOrderTraversal(0, self.result)
                
    return self.result

  def postOrder(self):
    self.result = []
    # Finish the implementation
    # You may need to add a new recursive method to do that
    def postOrderTraversal(i, result):
        if self.left[i] != -1:
            postOrderTraversal(self.left[i], result)        
        if self.right[i] != -1:
            postOrderTraversal(self.right[i], result)
        result.append(self.key[i])
    postOrderTraversal(0, self.result)
                
    return self.result

def main():
	tree = TreeOrders()
	tree.read()
	print(" ".join(str(x) for x in tree.inOrder()))
	print(" ".join(str(x) for x in tree.preOrder()))
	print(" ".join(str(x) for x in tree.postOrder()))

threading.Thread(target=main).start()

File: 2 Data Structures/Week 6 Binary Search Trees 2/Programming-Assignment-4/tree_orders/tree-orders.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#if defined(__unix__) || defined(__APPLE__)
#include <sys/resource.h>
#endif

using std::vector;
using std::ios_base;
using std::cin;
using std::cout;

using namespace std;

class TreeOrders {
  int n;
  vector <int> key;
  vector <int> left;
  vector <int> right;

public:
  void read() {
    cin >> n;
    key.resize(n);
    left.resize(n);
    right.resize(n);
    for (int i = 0; i < n; i++) {
      cin >> key[i] >> left[i] >> right[i];
    }
  }
  
  void inOrderTraversal(int i, vector<int> &result){
      if (i == -1) return;
      inOrderTraversal(left[i], result);
      result.push_back(key[i]);
      inOrderTraversal(right[i], result);
  }

  vector <int> in_order() {
    vector<int> result;
    // Finish the implementation
    // You may need to add a new recursive method to do that
    inOrderTraversal(0, result);

    return result;
  }
  
  void preOrderTraversal(int i, vector<int> &result){
      if (i == -1) return;
      result.push_back(key[i]);
      preOrderTraversal(left[i], result);
      preOrderTraversal(right[i], result);
  }

  vector <int> pre_order() {
    vector<int> result;    
    // Finish the implementation
    // You may need to add a new recursive method to do that
    preOrderTraversal(0, result);
    
    return result;
  }
  
  void postOrderTraversal(int i, vector<int> &result){
      if (i == -1) return;
      postOrderTraversal(left[i], result);
      postOrderTraversal(right[i], result);
      result.push_back(key[i]);
  }

  vector <int> post_order() {
    vector<int> result;
    // Finish the implementation
    // You may need to add a new recursive method to do that
    postOrderTraversal(0, result);
    
    return result;
  }
};

void print(vector <int> a) {
  for (size_t i = 0; i < a.size(); i++) {
    if (i > 0) {
      cout << ' ';
    }
    cout << a[i];
  }
  cout << '\n';
}

int main_with_large_stack_space() {
  ios_base::sync_with_stdio(0);
  TreeOrders t;
  t.read();
  print(t.in_order());
  print(t.pre_order());
  print(t.post_order());
  return 0;
}

int main (int argc, char **argv)
{
#if defined(__unix__) || defined(__APPLE__)
  // Allow larger stack space
  const rlim_t kStackSize = 16 * 1024 * 1024;   // min stack size = 16 MB
  struct rlimit rl;
  int result;

  result = getrlimit(RLIMIT_STACK, &rl);
  if (result == 0)
  {
      if (rl.rlim_cur < kStackSize)
      {
          rl.rlim_cur = kStackSize;
          result = setrlimit(RLIMIT_STACK, &rl);
          if (result != 0)
          {
              std::cerr << "setrlimit returned result = " << result << std::endl;
          }
      }
  }
#endif

  return main_with_large_stack_space();
}

